Adding documentation for Kanister Function KubeOps & Wait (#1153)
kanisterio/kanister: docs/functions.rst

   :align: left
   :widths: 5,5,5,15

   `operation`, Yes, `string`, ``create`` or ``delete`` Kubernetes resource
   `namespace`, No, `string`, namespace in which the operation is executed
   `spec`, No, `string`, resource spec that needs to be created
   `objectReference`, No, `map[string]interface{}`, object reference for delete operation

Example:

.. code-block:: yaml
  :linenos:

  - func: KubeOps
    name: createDeploy
    args:
      operation: create
      namespace: "{{ .Deployment.Namespace }}"
      spec:
        apiVersion: apps/v1
        kind: Deployment
        metadata:
          name: "{{ .Deployment.Name }}"
        spec:
          replicas: 1
          selector:
            matchLabels:
              app: example
          template:
            metadata:
              labels:
                app: example
            spec:
              containers:
              - image: busybox
                imagePullPolicy: IfNotPresent
                name: container
                ports:
                - containerPort: 80
                  name: http
                  protocol: TCP

Wait
----

This function is used to wait on a Kubernetes resource

> until desired state is reached.

Reviewer: ```suggestion
until a desired state is reached.
```

Arguments:

.. csv-table::
   :header: "Argument", "Required", "Type", "Description"
   :align: left
   :widths: 5,5,5,15

   `timeout`, Yes, `string`, wait timeout
   `conditions`, Yes, `map[string]interface{}`, keys should be ``allOf`` and/or ``anyOf`` with value as ``[]Condition``

``Condition`` struct:

.. code-block:: yaml
  :linenos:

  condition: "Go template condition that returns true or false"
  objectReference:
    apiVersion: "Kubernetes resource API version"
    resource: "Type of resource to wait for"
    name: "Name of the resource"

.. note::

>     We can refer object key-value in Go template condition with the help of a ``$`` prefix JSON-path syntax.

Reviewer: ```suggestion
    We can refer to the object key-value in Go template condition with the help of a ``$`` prefix JSON-path syntax.
```

      args:
        snapshotID: "{{ .ArtifactsIn.backupInfo.KeyValue.snapshotID }}"

KubeOps
-------

> This function is used to manage Kubernetes resources. We can use it to

Reviewer 2: instead of manage can we just use create/delete in the first statement itself?

Author: I mentioned manage because more operations will be added to this function

Author: If mentioning create/delete specifically sounds more correct then I'll make this change.

either create or delete resources on Kubernetes cluster.

Arguments:

.. csv-table::
   :header: "Argument", "Required", "Type", "Description"
   :align: left
   :widths: 5,5,5,15

>    `operation`, Yes, `string`, ``create`` or ``delete`` kubernetes resource

Reviewer 2: ```suggestion
   `operation`, Yes, `string`, ``create`` or ``delete`` Kubernetes resource
```

.. csv-table::
   :header: "Argument", "Required", "Type", "Description"
   :align: left
   :widths: 5,5,5,15

   `operation`, Yes, `string`, ``create`` or ``delete`` Kubernetes resource
   `namespace`, No, `string`, namespace in which the operation is executed
   `spec`, No, `string`, resource spec that needs to be created
   `objectReference`, No, `map[string]interface{}`, object reference for delete operation

Example:

.. code-block:: yaml
  :linenos:

  - func: KubeOps
    name: createDeploy
    args:
      operation: create
      namespace: "{{ .Deployment.Namespace }}"
      spec:
        apiVersion: apps/v1
        kind: Deployment
        metadata:
          name: "{{ .Deployment.Name }}"
        spec:
          replicas: 1
          selector:
            matchLabels:
              app: example
          template:
            metadata:
              labels:
                app: example
            spec:
              containers:
              - image: busybox
                imagePullPolicy: IfNotPresent
                name: container
                ports:
                - containerPort: 80
                  name: http
                  protocol: TCP

Wait
----

> This function is used to add wait on a Kubernetes resource
> till it matches a desired state.

Reviewer: ```suggestion
This function is used to wait on a Kubernetes resource
until a desired state is reached.
```

Arguments:

.. csv-table::
   :header: "Argument", "Required", "Type", "Description"
   :align: left
   :widths: 5,5,5,15

   `timeout`, Yes, `string`, wait timeout

>    `conditions`, Yes, `map[string]interface{}`, list of wait conditions that match ``anyOf`` or ``allOf`` the conditions

Reviewer 3: We can make it more readable and easy to consume, something like

```
conditions -> one of allOf or anyOf 
allOf/anyOf -> list of Condition ([]Conditio)

Condition:
args:
  condition: Go template condition to return tre or false
  objectReference: 
```

Let's also add a note about if someone wants to refer object key value in condition go template, they can fetch it using jsonpath syntax with prefix `$`

Example:

.. code-block:: yaml
  :linenos:

>   - func: Wait

Reviewer: Something isn't right with this block
```
Warning, treated as error:
/repo/docs/functions.rst:1241:Could not lex literal_block as "yaml". Highlighting skipped.
```
